### ISSUE TITLE
How to keep the number of mesh vertices during training in each epoch?

### ISSUE THREAD

**Issue by zhejia01** [SonSang/dmesh#12]
When training in multiview reconstruction tasks, we found that you provided 3 different method to initialize points. We tried 'sample' init_method and sample 3000 points on GT mesh. However, it seems that at the end of epoch_0 the number of mesh vertices decrease to less than 1500 in most cases. Even if we decrease the hyperparameter weight_regularizer_coef to 1e-9, the number of vertices at the end of epoch_0 phase_0 is still small.
Here is the curve of mesh vertices during multiview reconstruction of 252119 data in thingi32 dataset during epoch_0 phase_0. How to reduce the number of points lost during optimization?
![923ad42262efc9d65e4ff7680e4e7b8](https://github.com/user-attachments/assets/f95c936e-f9b5-4715-a54f-70574d25abe6)

**Comment by zhejia01** [comment]
A more extreme case is that if we sample 30000 points to initialize, the number of points left after optimization is less than 3000 for the case metioned above.

**Comment by SonSang** [comment]
Hi, thanks for pointing this out and sorry for the late reply.

As you pointed out, currently we can observe this kind of undesirable phenomenon when it comes to the multi-view reconstruction task. We believe this is because of the false gradients generated by merging two different renderers in the first phase of the optimization. As described in the paper, we use two different kinds of renderers in the first phase of the optimization, because one of the renderers does not generate visibility gradients. We use nvdiffrast's implementation to get those gradients, but there is possibility that naively merging these two renderers could end up in this kind of phenomenon.

One possible way to overcome is developing a genuine renderer that supports efficient rendering of transparent triangles and generates visibility gradients at the same time seamlessly, but we found out that it is too massive task to do it in this work and out of scope.

+) Or, you can simply turn off optimization for the per-point weights. In that way, no points will be lost during optimization.

**Comment by zhejia01** [comment]
> Hi, thanks for pointing this out and sorry for the late reply.
> 
> As you pointed out, currently we can observe this kind of undesirable phenomenon when it comes to the multi-view reconstruction task. We believe this is because of the false gradients generated by merging two different renderers in the first phase of the optimization. As described in the paper, we use two different kinds of renderers in the first phase of the optimization, because one of the renderers does not generate visibility gradients. We use nvdiffrast's implementation to get those gradients, but there is possibility that naively merging these two renderers could end up in this kind of phenomenon.
> 
> One possible way to overcome is developing a genuine renderer that supports efficient rendering of transparent triangles and generates visibility gradients at the same time seamlessly, but we found out that it is too massive task to do it in this work and out of scope.
> 
> +) Or, you can simply turn off optimization for the per-point weights. In that way, no points will be lost during optimization.

Thanks for your reply! However, we attempt to freeze the weight of points and successfully optimize our data in the first 200 steps but found cuda out of memory error after step 200 on a single RTX3090 24GB. It seems that the number of soft faces(more than 1 million) increases strictly monotonically with the increase of step, which might cause insufficient cuda memory errors during optimization. Would you please help me understand why the number of soft faces keeps increasing during optimization with frozen points weight?

**Comment by SonSang** [comment]
Hmm, it's interesting. Can you please check if the number of soft faces is exactly reason why that the out of memory error occurs? Because I think it is hardly possible for the number of soft faces causes that kind of error (even if there is 1 million faces, the memory is not so large). And in our current implementation, the number of soft faces increase monotonically because we keep appending the new possible face combinations to the already existing list. How about using the smaller number of seed points to avoid this issue?

**Comment by zhejia01** [comment]
> Hmm, it's interesting. Can you please check if the number of soft faces is exactly reason why that the out of memory error occurs? Because I think it is hardly possible for the number of soft faces causes that kind of error (even if there is 1 million faces, the memory is not so large). And in our current implementation, the number of soft faces increase monotonically because we keep appending the new possible face combinations to the already existing list. How about using the smaller number of seed points to avoid this issue?

Thanks for your answer！
